src/iceberg/sort_order.cc

```diff
@@ -80,4 +86,66 @@ bool SortOrder::Equals(const SortOrder& other) const {
   return order_id_ == other.order_id_ && fields_ == other.fields_;
 }
 
+bool SortOrder::IsBoundToSchema(const Schema& schema) const {
+  for (const auto& field : fields_) {
+    auto schema_field = schema.FindFieldById(field.source_id());
+    if (!schema_field.has_value() || schema_field.value() == std::nullopt) {
+      return false;
+    }
+
+    const auto& source_type = schema_field.value().value().get().type();
+    if (!source_type->is_primitive()) {
+      return false;
+    }
+
+    auto result = field.transform()->ResultType(source_type);
+    if (!result) {
+      return false;
+    }
+  }
+  return true;
+}
+
+Result<std::unique_ptr<SortOrder>> SortOrder::Make(const Schema& schema, int32_t sort_id,
+                                                   std::vector<SortField> fields) {
+  if (!fields.empty() && sort_id == kUnsortedOrderId) [[unlikely]] {
+    return InvalidArgument("{} is reserved for unsorted sort order", kUnsortedOrderId);
+  }
+
+  if (fields.empty() && sort_id != kUnsortedOrderId) [[unlikely]] {
+    return InvalidArgument("Sort order must have at least one sort field");
+  }
+
+  for (const auto& field : fields) {
+    ICEBERG_ASSIGN_OR_RAISE(auto schema_field, schema.FindFieldById(field.source_id()));
+    if (schema_field == std::nullopt) {
+      return InvalidArgument("Cannot find source column for sort field: {}", field);
+    }
+
+    const auto& source_type = schema_field.value().get().type();
+
+    if (!source_type->is_primitive()) {
+      return InvalidArgument("Cannot sort by non-primitive source field: {}",
+                             *source_type);
+    }
+
```

Comment:
```suggestion
    const auto& source_type = schema_field.value().get().type();
```

I think the check is covered by line below?